Prevent local relative downloadable products to be treated as "absolute"

artpi opened the pull request:
This PR has no functional change. It will strengthen security by making sure downloadable file paths are properly recognized.

## Motivation

Downloadable products have 3 "types":

- shortcodes that evaluate to a shortcode
- `relative` which correspond to relative location on a file disk
- `absolute` which begin with `//` and are meant to represent remote location of a file in the web.

There is however a way to sneak in `absolute` product via `///`, like `///home/server/public_html/wp-config.php`. This way also bypasses the mime type check, because this check is done only for `relative` products.
This PR strips `///` and replaces with `/` for these cases and prevents file paths with `///` to be treated as absolute ones.

Additionally, since file paths beginning with `///` will now be treated as relative ones, they will also be subject to `woocommerce_downloadable_file_exists` filter. We use that filter in D54789-code to prevent one user retrieving file path from another user's site.


## Q&A

**Is this a security problem?**

Using a file path with `///` in a downloadable product allows for bypassing allowed_mime_types and creating a download with `wp-config.php`. However this will not be downloadable - read along

**Will this break existing functionality?**

Maybe somebody wants to create a download with wp-config? This was a major consideration. I am convinced this is a bug, since class-wc-downloads.php already prevents from using `///`

https://github.com/woocommerce/woocommerce/blob/3aee9f2dbd1da2d4505aa23f77bd08561253c3c2/includes/class-wc-download-handler.php#L276


So currently it's posisble to create a downloadable file with wp-config, but not possible to download it.

**Will this break existing sites?**

If somebody would be very determined to provide a download for wp-config, they could hook in to allow to download it.
In that case, this PR only prevents adding new files with incorrect mime type. All the files previously configured will continue to work.


## Testing instructions
(Added by @peterfabian )
1. Create a downloadable product, add files that refer to 
  - absolute web address (e.g. https://via.placeholder.com/300/09f/fff.png)
  - file picked from media library
  - relative file (relative to plugin folder)
  - file starting with two slashes // that is considered a remote absolute path
  - file starting with 3 slashes that is considered a local absolute path
2. Buy the product and try to download all the files, see which work and which don't.
2. Update WC.
3. Try to download all the files again. All should work, except for local file download via 3 slashes---that should not be possible.

### Changelog entry

> Enhancement - Make sure downloadable file paths are properly recognized for strengthened security.

artpi committed:
Fix to treat absolute local paths as relative ones

peterfabian reviewed (commented):
Hi @artpi , thanks for your PR and apologies for the long wait. 

I tried exercising this code a bit and adding gradually more slashes (input as you can see below) and managed to get it down to 3 slashes on the output (after reload) when giving it 7 slashes on input, but I can't reproduce it now, so it might have been a fluke.

![Screenshot 2021-02-16 at 00 15 10](https://user-images.githubusercontent.com/2207451/108001664-5e012e00-6fed-11eb-9a7c-7d5dcc4330eb.png)


Could be worth adding a couple of test cases to `tests/legacy/unit-tests/product/class-wc-tests-product-download.php` if you have the time. Please let us know, otherwise, we can help out.

artpi committed:
Add a few tests

artpi committed:
Better url parsing

artpi committed:
Extra protections

artpi committed:
Fix tests

artpi commented:
Thank you @peterfabian !
> I tried exercising this code a bit and adding gradually more slashes (input as you can see below) and managed to get it down to 3 slashes on the output (after reload) when giving it 7 slashes on input, but I can't reproduce it now, so it might have been a fluke.

You are right! The way it was sanitizing the prepending slashes may have caused that. I replaced that with *sic* regex

> Could be worth adding a couple of test cases to tests/legacy/unit-tests/product/class-wc-tests-product-download.php

Good idea! Added.

peterfabian reviewed (commented):
Looks good now, thanks @artpi for the update and adding those tests! 
I have just a small nitpicky question regarding the regular expression, but otherwise, it's working great and you have my pre-approval.

artpi commented on includes/class-wc-product-download.php:
```diff
@@ -164,6 +172,11 @@ public function set_previous_hash( $value ) {
 	public function set_file( $value ) {
+		// A `///` is recognized as an "absolute", but on the filesystem, so it bypasses the mime check in `self::is_allowed_filetype`.
+		// This will strip extra prepending / to the maximum of 2.
+		if ( preg_match( '#^//+(/[^/].+)$#i', $value, $matches ) ) {
```
You are correct!
However, depending on greediness is a risky thing. I like to make my regexes as explicit as possible to protect against at least SOME issues :D

peterfabian replied:
afaik most regexp engines are by default greedy, so not assuming greediness might actually be problematic. There are some regexp engines that don't even support non-greedy matching. But it's only a minor thing, so thanks for the consideration.

peterfabian reviewed (approved):
Looks good, thanks for your help 👍

peterfabian commented:
I'm marking it as high impact just to highlight the fact we're changing how downloadable files work and that it might potentially create some disturbance to users.
